Public/Src/Cache/ContentStore/Library/Service/Grpc/BufferedReadStream.cs

```diff
@@ -0,0 +1,110 @@
+
+        private byte[] _storage; // buffer containing the next bytes to be read
+        private int _readPointer; // the next index to be read from the storage buffer
+
+        private int _position; // total bytes that have been read
+        private int _length; // total bytes that have been ingested
+
+        public override bool CanRead => true;
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!(buffer is null));
+            Debug.Assert(offset >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(count <= buffer.Length - offset);
+
+            int writePointer = offset;
+            int totalCount = 0;
+            while (totalCount < count)
+            {
+                Debug.Assert(_readPointer >= 0);
+                Debug.Assert((_storage is null) || (_readPointer < _storage.Length));
+
+                Debug.Assert(writePointer >= 0);
+                Debug.Assert(writePointer < buffer.Length);
+
+                if (_storage is null)
+                {
+                    _storage = await _reader().ConfigureAwait(false);
+                    if (_storage is null)
+                    {
+                        return totalCount;
+                    }
+                    _readPointer = 0;
+                    _length += _storage.Length;
+                }
+
+                Debug.Assert(!(_storage is null));
+
+                int copyCount = Math.Min(count - totalCount, _storage.Length - _readPointer);
+                Array.Copy(_storage, _readPointer, buffer, writePointer, copyCount);
+                _readPointer += copyCount;
+                writePointer += copyCount;
+                totalCount += copyCount;
+                _position += copyCount;
+
+                Debug.Assert(_readPointer <= _storage.Length);
+
+                if (_readPointer == _storage.Length)
+                {
+                    _storage = null;
+                    _readPointer = 0;
+                }
+
+            }
+
+            Debug.Assert(totalCount == count);
+            return totalCount;
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return ReadAsync(buffer, offset, count).Result;
+        }
+
+        public override bool CanWrite => false;
+
+        public override void Write(byte[] buffer, int offset, int count) => throw new InvalidOperationException();
+
+        public override bool CanSeek => false;
+
+        public override long Seek(long offset, SeekOrigin origin) => throw new InvalidOperationException();
+
+        public override void SetLength(long value) => throw new NotImplementedException();
+
+        public override Task FlushAsync(CancellationToken cancellationToken) => Task.CompletedTask;
+
+        public override void Flush() { }
+
+        public override long Length => _length;

```

Comment:
>_length [](start = 39, length = 7)

To me this seems unexpected behavior if I were using this class.
My understanding is that the Length property of a stream gives how much you can read and use that as a guide to read from the stream.
In this implementation it seems that we are updating the length as we are reading, so the other way around...

Reply:
Agree this is unconventional. We can't return conventional Length b/c we don't have it until we read to end. Idea here was to keep track of how much we had ingested. (This is almost but not quite Position; since we have an internal buffer we may have ingested a bit more than has actually been read.) I used this value in debugging but it is never called by the production code. I could (a) leave as is, (b) give this property a different name (and throw NotSupported for Length), or (c) not expose this quantity at all (and throw NotSupported for Length). Preference?